### PR TITLE
Restart stuck tasks

### PR DESCRIPTION
**Pull request by gabrielfs7**
Reeschedule stuck tasks

How to test:

1 - Have an advanced search env working with RDS worker:
2 - Create a orphan task and a stuck task with the following query:
```
-- Clear

DELETE FROM tq_task_log WHERE id NOT IN ('something');
DELETE FROM tq_indexation_queue WHERE id NOT IN (9999999999);

-- orphan

insert into tq_task_log (id, parent_id, task_name, parameters, label, status, master_status, owner, report, created_at, updated_at)
values (
           'https://advanced-search-tao.docker.localhost/ontologies/tao.rdf#i60f025be842ce119a5ed3e3b54d2fb5a',
           NULL,
           'oat\taoAdvancedSearch\model\Resource\Task\ResourceMigrationTask',
           '{"start":"0","classUri":"http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#Item","chunkSize":50,"pickSize":0,"repeat":true}',
           'Unit processing by oat\taoAdvancedSearch\model\Resource\Task\ResourceMigrationTask with chunk size 50',
           'enqueued',
           false,
           '',
           '{"type":"info","message":"Running task https:\/\/advanced-search-tao.docker.localhost\/ontologies\/tao.rdf#i60f025be842ce119a5ed3e3b54d2fb5a","data":null,"children":[{"type":"info","message":"Statement Migration Task","data":null,"children":[{"type":"info","message":"Respawning additional task","data":null,"children":[]}]}]}',
           '2021-07-15 12:10:38',
           '2021-07-15 12:10:44'
       );

-- with queue

insert into tq_task_log (id, parent_id, task_name, parameters, label, status, master_status, owner, report, created_at, updated_at)
values (
           'https://advanced-search-tao.docker.localhost/ontologies/tao.rdf#i60f025be842ce119a5ed3e3b54d2fb5c',
           NULL,
           'oat\taoAdvancedSearch\model\Resource\Task\ResourceMigrationTask',
           '{"start":"0","classUri":"http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#Item","chunkSize":50,"pickSize":0,"repeat":true}',
           'Unit processing by oat\taoAdvancedSearch\model\Resource\Task\ResourceMigrationTask with chunk size 50',
           'enqueued',
           false,
           '',
           '{"type":"info","message":"Running task https:\/\/advanced-search-tao.docker.localhost\/ontologies\/tao.rdf#i60f025be842ce119a5ed3e3b54d2fb5c","data":null,"children":[{"type":"info","message":"Statement Migration Task","data":null,"children":[{"type":"info","message":"Respawning additional task","data":null,"children":[]}]}]}',
           '2021-07-15 12:10:38',
           '2021-07-15 12:10:44'
       );

insert into tq_indexation_queue (id, message, visible, created_at)
values (
           285,
           '{"taskFqcn":"oat\\tao\\model\\taskQueue\\Task\\CallbackTask","metadata":{"__id__":"https:\/\/advanced-search-tao.docker.localhost\/ontologies\/tao.rdf#i60f025be842ce119a5ed3e3b54d2fb5c","__created_at__":"2021-07-15T12:10:38+00:00","__owner__":"","__master_status__":0,"__label__":"Unit processing by oat\\taoAdvancedSearch\\model\\Resource\\Task\\ResourceMigrationTask with chunk size 50","__callable__":"oat\\taoAdvancedSearch\\model\\Resource\\Task\\ResourceMigrationTask"},"parameters":{"start":"0","classUri":"http:\/\/www.tao.lu\/Ontologies\/TAOItem.rdf#Item","chunkSize":50,"pickSize":0,"repeat":true}}',
           false,
           '2021-07-15 12:10:38'
       );
```
3 - Run the new commands described in the README.md. Example:

```
sudo -u www-data php index.php 'oat\taoTaskQueue\scripts\tools\StuckTaskRestart' --queue indexation_queue --whitelist "oat\tao\model\search\tasks\UpdateResourceInIndex,oat\tao\model\search\tasks\UpdateClassInIndex,oat\tao\model\search\tasks\DeleteIndexProperty,oat\tao\model\search\tasks\RenameIndexProperties,oat\tao\model\search\tasks\UpdateDataAccessControlInIndex,oat\tao\model\search\tasks\AddSearchIndexFromArray,oat\taoAdvancedSearch\model\Resource\Task\ResourceMigrationTask,oat\taoAdvancedSearch\model\DeliveryResult\Service\DeliveryResultMigrationTask,oat\taoAdvancedSearch\model\Metadata\Task\MetadataMigrationTask" --age 300
```

and

```
sudo -u www-data php index.php 'oat\taoTaskQueue\scripts\tools\StuckTaskSummary' --queue indexation_queue --whitelist "oat\tao\model\search\tasks\UpdateResourceInIndex,oat\tao\model\search\tasks\UpdateClassInIndex,oat\tao\model\search\tasks\DeleteIndexProperty,oat\tao\model\search\tasks\RenameIndexProperties,oat\tao\model\search\tasks\UpdateDataAccessControlInIndex,oat\tao\model\search\tasks\AddSearchIndexFromArray,oat\taoAdvancedSearch\model\Resource\Task\ResourceMigrationTask,oat\taoAdvancedSearch\model\DeliveryResult\Service\DeliveryResultMigrationTask,oat\taoAdvancedSearch\model\Metadata\Task\MetadataMigrationTask" --age 300
```

https://oat-sa.atlassian.net/browse/ADF-514

### CODE REVIEW

File: test/unit/model/Repository/StuckTaskCollectionTest.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $task = $this->createMock(StuckTask::class);
+
+        $collection = new StuckTaskCollection(...[]);
+        $collection->add($task);
+
+        $this->assertSame($task, $collection->offsetGet(0));
+        $this->assertSame(1, $collection->count());
+    }
+}
```
Comment (boajer) [inline test/unit/model/Repository/StuckTaskCollectionTest.php:41]:
```suggestion
}

```

File: model/QueueBroker/RdsQueueBroker.php [hunk 2]
```diff
@@ -233,6 +235,49 @@ protected function doDelete($id, array $logContext = [])
         }
     }
 
+    public function getTaskByTaskLogId(string $taskLogId): ?CallbackTaskDecorator
```
Comment (boajer) [inline model/QueueBroker/RdsQueueBroker.php:241]: ~What about other classes inhered from `AbstractQueueBroker`?  If we can't go with a proper implementation may be at least something like `throw new NotSupportedMenthodExpeption()`  to maintain integrity ?~ Nevermind, https://github.com/oat-sa/extension-tao-task-queue/pull/141/files#diff-8a6c5dec1e176f9254a467381cf755030d818a2b290cba25de1c27730302db57R44 if you are checking in runtime it should be safe enougth

File: model/Repository/StuckTaskQuery.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        $this->statuses = $statuses;
+
+        /**
+         * Simulating approach provided at common_persistence_sql_Platform::getNowExpression()
+         * This is how we create new taskLogs in the system.
+         *
+         * @TODO Refactor this when we will have support for other Queue and TaskLog broker
+         */
+        $this->ageDateTime = new DateTime('now', new DateTimeZone('UTC'));
+        $this->ageDateTime->modify(sprintf('-%s seconds', $this->age));
```
Comment (boajer) [inline model/Repository/StuckTaskQuery.php]: Can `DateTimeImmutable` be considered here?

File: model/QueueBroker/RdsQueueBroker.php (hunk 2 above)
Comment (gabrielfs7) [reply model/QueueBroker/RdsQueueBroker.php:241]: The other implementations like SQS do not work with ids. This implementation is really to support RDS. I will open a tech debt task for this after we validate the solution.

File: model/Repository/StuckTaskQuery.php (hunk 3 above)
Comment (gabrielfs7) [reply model/Repository/StuckTaskQuery.php]: Can be, changed!

**Review by shpran — APPROVED** [review]
Well done!

- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by chinnu-focaloid — APPROVED** [review]
- [X] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

File: model/QueueBroker/RdsQueueBroker.php (hunk 2 above)
Comment (gabrielfs7) [reply model/QueueBroker/RdsQueueBroker.php:241]: Tech debt here: https://oat-sa.atlassian.net/browse/ADF-556

**Review by boajer — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Comment by AntonNeviarouski** [comment]
QA verified on http://test-advanced-criteria.playground.kitchen.it.taocloud.org:42161/

**Pull request merged**